### PR TITLE
Add Workflow, WorkflowCard, and WorkflowActivity

### PR DESCRIPTION
**Pull request by jessejanderson**
Still in progress, but tests appear to be working for Workflow Cards and Workflow Card Activities (3 level deep!)

I went on the assumption that there will always be _some_ link available, either the specific one I want or at least a `self`. That way I can always get the preceding IDs in some manner.

I'm not sure if there's a way to capture the `meta` info from the json, but that holds a `parent id` if we didn't want to parse the `self` url.

### CODE REVIEW

File: lib/pco_api/people/workflow/card.ex [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  linked_association :activities
+  linked_association :notes
+
+  @doc """
+  Gets associated Workflow.Card records from a Workflow Record from links.
+
+  ## Example:
+
+      iex> %PcoApi.Record{type: "Person", links: %{"addresses" => "http://example.com"}} |> PcoApi.People.Address.get
+      %PcoApi.Record{type: "Address", ...}
```
Comment (geolessel) [inline lib/pco_api/people/workflow/card.ex]: You're gonna want to update the docs from addresses to workflows.

File: lib/pco_api/people/workflow/card/activity.ex [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  Since an Activity is always associated with a Workflow Card in Planning Center Online,
+  a Record of type "WorkflowCard" is required in order to retrieve that Workflow Card's
+  associated Activities.
+  """
+
+  use PcoApi.Actions
+  endpoint "people/v2/workflows/"
+
+  @doc """
+  Gets associated Workflow.Card records from a Workflow Record from links.
```
Comment (geolessel) [inline lib/pco_api/people/workflow/card/activity.ex]: Needs a space after the period.

File: lib/pco_api/people/workflow/card/activity.ex [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+      iex> %PcoApi.Record{type: "Person", id: 1} |> PcoApi.People.Address.get
+      %PcoApi.Record{type: "Address", id: 1, ...}
+
+  """
+  # def get(%PcoApi.Record{type: "Person", id: id}), do: get("people/#{id}/addresses")
+  def get(%PcoApi.Record{type: "WorkflowCard", id: id, links: %{"self" => url}}) do
+    workflow_id = url |> get_workflow_id
+    get("#{workflow_id}/cards/#{id}/activities")
+  end
```
Comment (geolessel) [inline lib/pco_api/people/workflow/card/activity.ex]: Could you also just plop `activities` on the end of `self`?

``` elixir
def get(%PcoApi.Record{type: "WorkflowCard", links: %{"self" => self}}, do: get self <> "/activities"
```

File: lib/pco_api/people/workflow/card/activity.ex [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+  def get(card, id) when is_integer(id), do: get(card, Integer.to_string(id))
+  def get(%PcoApi.Record{type: "WorkflowCard", links: %{"self" => url}}, id) do
+    get(url <> "/activities/" <> id)
+  end
+
+  defp get_workflow_id(url) do
+    [_, workflow_id] = Regex.run(~r/workflows\/(.*)\/cards/, url)
+    workflow_id
+  end
+end
```
Comment (geolessel) [inline lib/pco_api/people/workflow/card/activity.ex:64]: This module also needs updates to inline docs.

File: test/fixtures/workflow_card.json [hunk 5]
```diff
@@ -0,0 +1 @@
...o_back","notes":"https://api.planningcenteronline.com/people/v2/workflows/1/cards/101/notes","person":"https://api.planningcenteronline.com/people/v2/people/6799543","promote":"https://api.planningcenteronline.com/people/v2/workflows/1/cards/101/promote","remove":"https://api.planningcenteronline.com/people/v2/workflows/1/cards/101/remove","restore":"https://api.planningcenteronline.com/people/v2/workflows/1/cards/101/restore","send_email":"https://api.planningcenteronline.com/people/v2/workflows/1/cards/101/send_email","skip_step":"https://api.planningcenteronline.com/people/v2/workflows/1/cards/101/skip_step","snooze":"https://api.planningcenteronline.com/people/v2/workflows/1/cards/101/snooze","tasks":"https://api.planningcenteronline.com/people/v2/workflows/1/cards/101/tasks","unsnooze":"https://api.planningcenteronline.com/people/v2/workflows/1/cards/101/unsnooze","self":"https://api.planningcenteronline.com/people/v2/workflows/1/cards/101"},"relationships":{"assignee":{"data":{"type":"Assignee","id":"7291209"}},"current_step":{"data":{"type":"CurrentStep","id":"40313"}}}},"included":[],"meta":{"can_include":["person","assignee","tasks"],"parent":{"id":"1","type":"Workflow"}}}
```
Comment (geolessel) [inline test/fixtures/workflow_card.json]: It looks like there are real people ids in here. You might want to remove.

File: test/fixtures/workflow_card_activities.json [hunk 6]
```diff
@@ -0,0 +1 @@
+{"links":{"self":"https://api.planningcenteronline.com/people/v2/workflows/14589/cards/152935/activities"},"data":[{"type":"WorkflowCardActivity","id":"1380904","attributes":{"comment":"created a new note","content":"A note.","created_at":"2016-04-20T20:53:46Z","person_avatar_url":"https://avatars.planningcenteronline.com/uploads/person/9061163-1433195003/avatar.1.png?g=68x68%23","person_name":"Shane Bonham","reassigned_to_avatar_url":null,"reassigned_to_name":null,"subject":null,"type":"note"},"links":{"self":"https://api.planningcenteronline.com/people/v2/workflows/14589/cards/152935/activities/1380904"}}],"included":[],"meta":{"total_count":1,"count":1,"can_order_by":["id"],"parent":{"id":"152935","type":"WorkflowCard"}}}
```
Comment (geolessel) [inline test/fixtures/workflow_card_activities.json]: Same here. Remove the actual IDs and names.

File: test/fixtures/workflow_card_activity.json [hunk 7]
```diff
@@ -0,0 +1 @@
+{"data":{"type":"WorkflowCardActivity","id":"1380904","attributes":{"comment":"created a new note","content":"A note.","created_at":"2016-04-20T20:53:46Z","person_avatar_url":"https://avatars.planningcenteronline.com/uploads/person/9061163-1433195003/avatar.1.png?g=68x68%23","person_name":"Shane Bonham","reassigned_to_avatar_url":null,"reassigned_to_name":null,"subject":null,"type":"note"},"links":{"self":"https://api.planningcenteronline.com/people/v2/workflows/14589/cards/152935/activities/1380904"}},"included":[],"meta":{"parent":{"id":"152935","type":"WorkflowCard"}}}
```
Comment (geolessel) [inline test/fixtures/workflow_card_activity.json]: Needs real info removed.

File: test/fixtures/workflow_card_list.json [hunk 8]
```diff
@@ -0,0 +1 @@
...le/v2/workflows/14589/cards/182833"},"relationships":{"assignee":{"data":{"type":"Assignee","id":"7291209"}},"current_step":{"data":{"type":"CurrentStep","id":"40313"}}}},{"type":"WorkflowCard","id":"153387","attributes":{"completed_at":null,"created_at":"2016-04-21T15:54:32Z","snooze_until":null,"updated_at":"2016-05-06T17:46:09Z"},"links":{"self":"https://api.planningcenteronline.com/people/v2/workflows/14589/cards/153387"},"relationships":{"assignee":{"data":{"type":"Assignee","id":"9061163"}},"current_step":{"data":{"type":"CurrentStep","id":"40313"}}}},{"type":"WorkflowCard","id":"156297","attributes":{"completed_at":null,"created_at":"2016-04-23T00:52:40Z","snooze_until":null,"updated_at":"2016-05-12T15:26:29Z"},"links":{"self":"https://api.planningcenteronline.com/people/v2/workflows/14589/cards/156297"},"relationships":{"assignee":{"data":{"type":"Assignee","id":"9061163"}},"current_step":{"data":{"type":"CurrentStep","id":"40313"}}}}],"included":[],"meta":{"total_count":4,"count":4,"can_order_by":["created_at","updated_at","completed_at"],"can_query_by":["assignee_id","step_id","stage"],"can_include":["person","assignee","tasks"],"parent":{"id":"14589","type":"Workflow"}}}
```
Comment (geolessel) [inline test/fixtures/workflow_card_list.json]: Needs real info removed.

File: test/fixtures/workflow_step_list.json [hunk 9]
```diff
@@ -0,0 +1 @@
+{"links":{"self":"https://api.planningcenteronline.com/people/v2/workflows/14589/steps"},"data":[{"type":"WorkflowStep","id":"47401","attributes":{"auto_snooze_days":null,"created_at":"2016-05-13T18:01:23Z","description":null,"my_ready_card_count":0,"name":"This step does have an assignee.","sequence":0,"total_ready_card_count":0,"total_snoozed_card_count":0,"updated_at":"2016-05-13T18:04:19Z"},"links":{"self":"https://api.planningcenteronline.com/people/v2/workflows/14589/steps/47401"}},{"type":"WorkflowStep","id":"40313","attributes":{"auto_snooze_days":null,"created_at":"2016-04-18T23:08:44Z","description":"What's more, the card in it doesn't have an assignee!","my_ready_card_count":0,"name":"This Step Has No Default Assignee!","sequence":1,"total_ready_card_count":6,"total_snoozed_card_count":0,"updated_at":"2016-05-13T18:04:19Z"},"links":{"self":"https://api.planningcenteronline.com/people/v2/workflows/14589/steps/40313"}}],"included":[],"meta":{"total_count":2,"count":2,"can_order_by":["name","created_at","updated_at","sequence"],"can_query_by":["name","created_at","updated_at"],"can_include":["default_assignee"],"parent":{"id":"14589","type":"Workflow"}}}
```
Comment (geolessel) [inline test/fixtures/workflow_step_list.json]: Remove real info.

File: test/pco_api/people/workflow_test.exs [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+  #   workflow |> Workflow.get
+  # end
+
+  # test ".get gets Workflow details even if no self link", %{bypass: bypass} do
+  #   Bypass.expect bypass, fn conn ->
+  #     assert "/people/v2/workflows/1000" == conn.request_path
+  #     Plug.Conn.resp(conn, 200, Fixture.read("workflow.json"))
+  #   end
+  #   %PcoApi.Record{id: "1000", links: %{}} |> Workflow.get
+  # end
```
Comment (geolessel) [inline test/pco_api/people/workflow_test.exs]: Remove commented code

**Comment by geolessel** [comment]
Still needs some massaging but looking good.

File: lib/pco_api/people/workflow/card.ex (hunk 1 above)
Comment (jessejanderson) [inline lib/pco_api/people/workflow/card.ex]: Hence the TODO at the top. Would've done done it before merge if it wasn't 1am.

File: lib/pco_api/people/workflow/card/activity.ex (hunk 4 above)
Comment (jessejanderson) [inline lib/pco_api/people/workflow/card/activity.ex:64]: Yep - TODO at the top.

File: lib/pco_api/people/workflow/card/activity.ex (hunk 3 above)
Comment (jessejanderson) [inline lib/pco_api/people/workflow/card/activity.ex]: Well sure, if you want to make it _easy_. 😉

**Comment by geolessel** [comment]
@jessejanderson why/how did the .beam files get added? Do we need those?

File: lib/pco_api/people/workflow/card.ex [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+      iex> %PcoApi.Record{type: "Workflow", links: %{"cards" => "http://example.com"}} |> Card.get
+      %PcoApi.Record{type: "WorkflowCard", ...}
+
+  """
+  def get(%PcoApi.Record{type: "Workflow", links: %{"cards" => url}}), do: get url
+
+  @doc """
+  Gets associated WorkflowCard records from a Workflow Record when no cards link is found.
+
+  Sometimes a record my not include an cards link. This function recreates a URL to
```
Comment (geolessel) [inline lib/pco_api/people/workflow/card.ex]: `my` -> "may"

`an` -> "a"

File: lib/pco_api/people/workflow/card/activity.ex [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+  Requires a WorkflowCard with an activities link and a WorkflowCardActivity Id.
+
+  ## Example:
+
+      iex> %PcoApi.Record{type: "WorkflowCard", id: 1, links: %{"activities" => "http://example.com"}} |> Activity.get(2)
+      %PcoApi.Record{type: "WorkflowCardActivity", id: 2} # for WorkflowCard.id == 1
+
+  """
+  def get(card, id) when is_integer(id), do: get(card, Integer.to_string(id))
```
Comment (geolessel) [inline lib/pco_api/people/workflow/card/activity.ex]: Let's pattern match on `card` too.

**Comment by geolessel** [comment]
Nice. 👍  If you could remove the .beam files before merging, I'd appreciate it.

**Pull request merged**